### ISSUE TITLE
Refactor/Rewrite of Commands module / Updater are ready for testing

### ISSUE THREAD

**Issue by MichaelLeeHobbs** [AJS-development/Ogar-unlimited#235]
This is a massive change and needs to be fully test to include every command. Additionally, the updated has been rewritten basically from scratch. There is still some minor work to be done on it, but it is basically done. Lastly, I will merge the master branch into dev later today. You do not need to wait on the merge before you can pull the dev to start testing.

Please post issues with the dev branch here.

**Comment by MichaelLeeHobbs** [comment]
CI build fails. Will take a look at it latter.

**Comment by Andrews54757** [comment]
Also, how is the commands rewritten? After I looked at the dev branch, I was confused as major things were changed. And how does it flow now? That means I have to relearn how to program commands

**Comment by Andrews54757** [comment]
Also I want the updater to know what files to update using the file from the github server, not in local files.js but in online files.js with a format of this
```
filedir downloadserverurl(optional)
filedir2 ....
```
I need to have the downloadserverurl option because some files can't be downloaded your way such as expiramental v2.js

**Comment by Andrews54757** [comment]
Never mind, I realized what you did but I still want the downloadserverurl option. Good Job!

**Comment by Andrews54757** [comment]
Nvm, you did include that option. But I have a question, what is the hash

**Comment by Andrews54757** [comment]
Actually though I was thinking of something more simple than files.json, I was thinking about a shorter, easier file with the format I said above.

So it downloads the file files.txt
Then it splits it on line breaks
Then it splits those splits into smaller splits by :  and one side is the dir the other the optional downloadserverurl.
If downloadserverurl is not specified, it uses the directory and the default raw.github.com thing

**Comment by MichaelLeeHobbs** [comment]
On the commands. They are each separated into their own file and then required in the commands.js file. If you take a look at it I think it will become quite clear. Not much really changed there other than sorting them into many small easy to maintain files vs have a nearly 3000 line long file.

On the updater. This is not the final long term project. While working through it and thinking about where we want to go with having multiple processes a number of idea's came to mind. But first let me cover the file format issue and what a hash is. A hash is a cryptographic tool that lets you determine if A = B. So if file A on github has the same hash as B on the local system then they are the same, ie up-to-date. Now, the reason I went with a JSON file, is because it is a self describing file system that is already built into node/js. It is a good easy way to take data and turn it into a file and vice-versa. I do realize they are harder to write than a simple text file but we shouldn't be writing that file anyway. I will eventually build a tool to generate it for us so we don't have to worry about it. I just ran out of time this weekend. I still need to merge the master and dev branch as I am positive it is going to have conflicts.

My thoughts on the updater in the future. After we get the main loop refactored and setup into processes we can take the next steps on the updater. So my idea is to have an initial process that starts up. Call in the controller/launcher process. The controller process ONLY monitors and starts other processes. This is key because it will allow us to have code that should basically never change. With that in place it will allow us to have the updater download new files and tell the controller we need to restart or even better, schedule a restart. The controller can then shut down all other processes except for it's self and then relaunch them. No more restart issues with Linux/OSX. Secondly, by doing this we can also have the controller relaunch the server when something crashes as the controller should never crash.

BTW, have you been able to test all of the commands? I'll feel much safer merging this over into master when someone else test them all.
